File: src/ai/flows/improve-recommendations-from-feedback.ts
'use server';
/**
 * @fileOverview A flow that improves recipe recommendations based on user feedback.
 *
 * - improveRecommendationsFromFeedback - A function that handles the process of improving recommendations based on feedback.
 * - ImproveRecommendationsFromFeedbackInput - The input type for the improveRecommendationsFromFeedback function.
 * - ImproveRecommendationsFromFeedbackOutput - The return type for the improveRecommendationsFromFeedback function.
 */

import {ai} from '@/ai/genkit';
import {z} from 'genkit';

const ImproveRecommendationsFromFeedbackInputSchema = z.object({
  recipeId: z.string().describe('The ID of the recipe the user is providing feedback for.'),
  feedbackType: z
    .enum(['upvote', 'downvote', 'skip'])
    .describe('The type of feedback the user is providing.'),
  userId: z.string().describe('The ID of the user providing feedback.'),
});
export type ImproveRecommendationsFromFeedbackInput = z.infer<
  typeof ImproveRecommendationsFromFeedbackInputSchema
>;

const ImproveRecommendationsFromFeedbackOutputSchema = z.object({
  success: z.boolean().describe('Whether the feedback was successfully processed.'),
  message: z.string().describe('A message indicating the outcome of the feedback processing.'),
});
export type ImproveRecommendationsFromFeedbackOutput = z.infer<
  typeof ImproveRecommendationsFromFeedbackOutputSchema
>;

export async function improveRecommendationsFromFeedback(
  input: ImproveRecommendationsFromFeedbackInput
): Promise<ImproveRecommendationsFromFeedbackOutput> {
  return improveRecommendationsFromFeedbackFlow(input);
}

const prompt = ai.definePrompt({
  name: 'improveRecommendationsFromFeedbackPrompt',
  input: {schema: ImproveRecommendationsFromFeedbackInputSchema},
  output: {schema: ImproveRecommendationsFromFeedbackOutputSchema},
  prompt: `You are an AI assistant that improves recipe recommendations based on user feedback.

  A user has provided the following feedback for a recipe:

  User ID: {{{userId}}}
  Recipe ID: {{{recipeId}}}
  Feedback Type: {{{feedbackType}}}

  Analyze the feedback and update your understanding of the user's preferences.
  Based on this feedback, adjust the algorithm to provide better recommendations in the future.

  Return a success boolean and message indicating the outcome of the feedback processing.
  The message should summarize how the feedback has been processed and what changes have been made to the recommendation algorithm.
  The output should be in JSON format.
  `, // Ensure output is in JSON format.
});

const improveRecommendationsFromFeedbackFlow = ai.defineFlow(
  {
    name: 'improveRecommendationsFromFeedbackFlow',
    inputSchema: ImproveRecommendationsFromFeedbackInputSchema,
    outputSchema: ImproveRecommendationsFromFeedbackOutputSchema,
  },
  async input => {
    const {output} = await prompt(input);
    return output!;
  }
);

File: backend/src/ai/flows/improve_recommendations_from_feedback.py
"""
Improve Recommendations from Feedback Flow
A flow that improves recipe recommendations based on user feedback.
"""
from pydantic import BaseModel, Field
from enum import Enum
from src.ai.genkit import ai


class FeedbackType(str, Enum):
    """Enum for feedback types."""
    UPVOTE = "upvote"
    DOWNVOTE = "downvote"
    SKIP = "skip"


class ImproveRecommendationsFromFeedbackOutput(BaseModel):
    """Output schema for feedback processing results."""
    success: bool = Field(description="Whether the feedback was successfully processed")
    message: str = Field(description="A message indicating the outcome of the feedback processing")


@ai.flow()
async def improve_recommendations_from_feedback(
    recipe_id: str,
    feedback_type: str,
    user_id: str
) -> ImproveRecommendationsFromFeedbackOutput:
    """
    Processes user feedback to improve future recipe recommendations.
    
    Args:
        recipe_id: The ID of the recipe the user is providing feedback for
        feedback_type: The type of feedback (upvote, downvote, or skip)
        user_id: The ID of the user providing feedback
        
    Returns:
        ImproveRecommendationsFromFeedbackOutput with success status and message
    """
    # Construct the prompt
    prompt = f"""You are an AI assistant that improves recipe recommendations based on user feedback.

A user has provided the following feedback for a recipe:

User ID: {user_id}
Recipe ID: {recipe_id}
Feedback Type: {feedback_type}

Analyze the feedback and update your understanding of the user's preferences.
Based on this feedback, adjust the algorithm to provide better recommendations in the future.

Provide a summary of:
- Whether the feedback was successfully processed (success: true/false)
- A message summarizing how the feedback has been processed and what changes have been made"""

    # Generate the response using ai.generate
    result = await ai.generate(
        prompt=prompt,
        output_schema=ImproveRecommendationsFromFeedbackOutput,
    )
    
    return result.output
